registry/handlers/app.go

```diff
@@ -1072,3 +1058,42 @@ func startUploadPurger(ctx context.Context, storageDriver storagedriver.StorageD
 		}
 	}()
 }
+
+// buildRegexValidationOptions returns a slice of registry options containing
+// rules for regex validation of a certain field (using allow/deny sections)
+func buildRegexValidationOptions(sectionName string, defaultAllowAll bool, allow []string, deny []string, allowOptFunc, denyOptFunc func(r *regexp.Regexp) storage.RegistryOption) []storage.RegistryOption {
+	var options []storage.RegistryOption
+	if len(allow) == 0 && len(deny) == 0 {
+		if defaultAllowAll {
+			options = append(options, storage.ManifestURLsAllowRegexp(regexp.MustCompile("^.*$"))) // allow everything by default
+		} else {
+			options = append(options, storage.ManifestURLsAllowRegexp(regexp.MustCompile("^$"))) // allow nothing by default
+		}
+	} else {
+		if len(allow) > 0 {
+			for i, s := range allow {
+				// Validate via compilation.
+				if _, err := regexp.Compile(s); err != nil {
+					panic(fmt.Sprintf("%s.allow: %s", sectionName, err))
+				}
+				// Wrap with non-capturing group.
+				allow[i] = fmt.Sprintf("(?:%s)", s)
+			}
+			re := regexp.MustCompile(strings.Join(allow, "|"))
+			options = append(options, allowOptFunc(re))
+		}
+		if len(deny) > 0 {
+			for i, s := range deny {
+				// Validate via compilation.
+				if _, err := regexp.Compile(s); err != nil {
+					panic(fmt.Sprintf("%s.deny: %s", sectionName, err))
+				}
+				// Wrap with non-capturing group.
+				deny[i] = fmt.Sprintf("(?:%s)", s)
+			}
+			re := regexp.MustCompile(strings.Join(deny, "|"))
```

Comment:
From a performance perspective; wondering if keeping the regexes as a slice would be better. Doing a (quite naive) benchmark (code below) it seems like looping over the regexes (allowing early matches if needed) could be faster;

```
BenchmarkAll
BenchmarkAll-8     	  627915	      1816 ns/op	       0 B/op	       0 allocs/op
BenchmarkSlice
BenchmarkSlice-8   	 2515844	       485.7 ns/op	       0 B/op	       0 allocs/op
```

<details>

```go
package main

import (
	"regexp"
	"strings"
	"testing"
)

var regexes = []string{
	`^application/vnd\.oci\.image\.layer\.(v|x)1\.tar$`,
	`^application/vnd\.oci\.image\.layer\.(v|x)2\.tar$`,
	`^application/vnd\.oci\.image\.layer\.(v|x)3\.tar$`,
	`^application/vnd\.oci\.image\.layer\.(v|x)4\.tar$`,
	`^application/vnd\.oci\.image\.layer\.(v|x)5\.tar$`,
	`^application/vnd\.oci\.image\.layer\.(v|x)6\.tar$`,
	`^application/vnd\.oci\.image\.layer\.(v|x)7\.tar$`,
}

var (
	all     string
	reAll   *regexp.Regexp
	reSlice []*regexp.Regexp
)

func init() {
	reAll = regexp.MustCompile(`(` + strings.Join(regexes, `)|(?:`) + `)`)

	reSlice = []*regexp.Regexp{}

	for _, re := range regexes {
		reSlice = append(reSlice, regexp.MustCompile(re))
	}

}

func BenchmarkAll(b *testing.B) {
	b.ReportAllocs()
	for i := 0; i < b.N; i++ {
		_ = reAll.MatchString("application/vnd.oci.image.layer.v7.tar")
	}
}

func BenchmarkSlice(b *testing.B) {
	b.ReportAllocs()
	for i := 0; i < b.N; i++ {
		for i := 0; i < len(reSlice); i++ {
			if reSlice[i].MatchString("application/vnd.oci.image.layer.v7.tar") {
				break
			}
		}
	}
}
```

</details>